### ISSUE TITLE
match-with? as matcher, instead of assert expression

### ISSUE THREAD

**Issue by lennartbuit** [nubank/matcher-combinators#129]
Hey, loving this library! One thing I just stumbled upon was wanting to `match?` with a nested structure, but asserting that it was exactly equals.

Now, this can be done with `match-with?` (clojure test user here). However, we usually write tests as long `are` blocks with many different cases exercising the same function. Sometimes that means that you need to special case a matcher, e.g. in the general case I need `m/embeds`, but in this particular case I need `m/equals`. With this particular combination, its a bit unfortunate that `match-with?` is an assert expression, as that means breaking out of the current `are`, writing a new `are` with `match-with?`, etc.

So I was thinking, would it be possible to expose `match-with?` not just as an assert expression, but also as a matcher (like `m/equals`), and if so, is that desirable :)?

**Comment by dchelimsky** [comment]
The nice thing about assert expressions is you just load a namespace and there they are. The bad thing about assert expressions is you just load a namespace and there they are.

I could see leaving the assert expressions in place, and adding `match-with?`, `match-equals?` and `match-roughly?` to the `standalone` namespace, but it would be on you to namespace them in an `is` expression. I need to get some feedback from other maintainers before pursuing this, and make no guarantees we'll do it at this point.

In the mean time, one alternative to breaking out of `are` _might_ (I may be missing context) be to add the matcher you want to use to the `are` arguments and then wrap it around the `are` predicate, e.g. 

```clojure
(are [matcher expected actual]
  (match? (matcher expected) actual)
  m/embeds expected1 actual
  m/equals expected2 actual
  ,,,)
```

Would that work for now?

**Comment by philomates** [comment]
Hey @lennartbuit, cool to hear you are enjoying the lib!
I agree that having the only way to set the default matchers at the outer `match?` level is a bit constraining. Technically speaking, it should be possible to introduce constructs that would recursively alter the default matchers for lexically scoped substructures. I'm not exactly sure why I didn't do that when I implemented `match-with?` but thinking about it now it feels hard to "get right" and is definitely a trade-off of power/expressiveness and readability/usability.

That said, if we can think of compelling use cases and a cohesive design, it would be cool to try!

Brainstorming what this might look like:
 - remove `match-with?`, `match-equals?`, etc and just leave `match?`
 - add "context modifiers" (horrible name), which are sorta like matchers, but apply recursively

```clojure
;; so
(match-equals?
  [{:a [{:b [1]}]}]
  whatever)

;; would become
(match?
  (equals-as-default [{:a [{:b [1]}]}])
  whatever)
```

with that, I would imagine one could achieve what I think you are trying to do (though I didn't completely follow)

```clojure
(match?
  (embeds-as-default [(equals-as-default {:a [{:b [1]}]})])
  whatever)
```

This sorta _feels_ scary to me because it seems messy given each "context modifier" applies to all subcontexts, and so you need to jump out to the closes modifier to know how the match semantics that a datastructure will default to. Currently you need to only look at the top level `match-*?` and the directly encompassing matcher.

Does this idea make any sense? Would this type of thing help your problem, or am I missing the point?

**Comment by lennartbuit** [comment]
Hey, thanks both for responding so swiftly. I think we are on the same line, @philomates. For completeness, I'll add an example of what I was trying to achieve. Can't do it with actual code (property of my employer), so I crafted an example.

Let's say that we have a function like this:

```clojure
(defn find-people
  [db {:keys [query order]}]
  (cond->> db
    query (filter (fn [{:keys [first-name last-name]}]
                    (or (str/includes? first-name query)
                        (str/includes? last-name query))))
    true  shuffle ;; Emulating 'arbitrary database order'
    order (sort-by order)))
```

With some example data, lets say like this[1]:

```clojure
(def example-db
  [{:first-name "Lennart"
    :last-name  "Buit"
    :location   {:country "The Netherlands"}
    :company    {:name "Nedap"}}
   {:first-name "David"
    :last-name  "Chelimsky"
    :location   {:country "United States"
                 :city    "Chicago"}
    :company    {:name "Cognitect"}}
   {:first-name "Phillip"
    :last-name  "Mates"
    :location   {:country "Germany"
                 :city    "Berlin"}
    :company    {:name "Nubank"}}])
```

What we like to do is write focussed test cases. We hope that this reduces the amount of unrelated failures (e.g. 50% of your test suite fell over because some unrelated property changed). In the case of this search function, that means that we have very lax matchers that mostly assert about _what elements_ are returned, but not _exactly what shape_. Matcher combinators aids us nicely here:

```clojure
(deftest find-people-test
  (are [input expected] (match? expected
                                (find-people example-db input))
    {:query "Lenn"}
    [{:first-name "Lennart"}]

    {:query "i"}
    (m/in-any-order
     [{:last-name "Buit"}
      {:first-name "David"
       :last-name  "Chelimsky"}
      {:first-name "Phillip"}])

    {:order :first-name}
    [{:first-name "David"}
     {:first-name "Lennart"}
     {:first-name "Phillip"}]))
```

But at some point, we may want to guarantee that search results have a specific shape. Let's say because we are exposing this as a REST endpoint. I have two options. I can either add a new case to my `are`, with a manually recursively applied `m/equals`:

```clojure
(deftest find-people-test
  (are [input expected] (match? expected
                                (find-people example-db input))
    ;; ... old test cases omitted for brevity ...
    {:query "Phillip"}
    [(m/equals
      {:first-name "Phillip"
       :last-name  "Mates"
       :location   (m/equals
                    {:country "Germany"
                     :city    "Berlin"})
       :company    (m/equals
                    {:name    "Nubank"})})]))
```

Or I can start a new `are`, and use `match-equals?`:

```clojure
(deftest find-people-test
  ;; ... previous `are` with `match?` omitted for brevity
  (are [input expected] (match-equals? expected
                                       (find-people example-db input))
    {:query "Phillip"}
    [{:first-name "Phillip"
      :last-name  "Mates"
      :location   {:country "Germany"
                   :city    "Berlin"}
      :company    {:name    "Nubank"}}]))
```

----

So indeed, I was looking for (what Phillip dubbed) a context modifier that _looks_ like one of those matchers, but recursivley. E.g. something like this [2]:

```clojure
(deftest find-people-test
  (are [input expected] (match? expected
                                (find-people example-db input))
    ;; ... old test cases omitted for brevity ...
    {:query "Phillip"}
    (c/with-equals [{:first-name "Phillip"
                     :last-name  "Mates"
                     :location   {:country "Germany"
                                  :city    "Berlin"}
                     :company    {:name    "Nubank"}}])))
```

I thought that would be a nice symmetry, I can change matchers in line, but I can't change them recursively inline. 

That said, I can understand that this is a very niche kinda thing. Maybe the answer here is 'cool idea, but way too much effort for what it is worth', or maybe the answer is, 'oh I saw this quite often, let's do something about it', I just wanted to share my use case :)!

-----

[1] I hope its okay that I took your public information as example :)
[2] With this obviously being pseudo code

**Comment by dchelimsky** [comment]
I've got a prototype implementation in PR #134. Feedback wanted!

**Comment by dchelimsky** [comment]
@lennartbuit with #134, your example would look like this:

```clojure
(deftest find-people-test
  (are [input expected] (match? expected
                                (find-people example-db input))
    ;; ... old test cases omitted for brevity ...
    {:query "Phillip"}
    [(m/match-with {clojure.lang.IPersistentMap m/equals}
      {:first-name "Phillip"
       :last-name  "Mates"
       :location   {:country "Germany"
                    :city    "Berlin"}
       :company    {:name    "Nubank"}})]))
```

It's slightly less slick than `c/with-equals`, but it has these benefits:

* aligns with `match-with?` API
* flexible (can apply to any type)
* intention revealing

Also, as you'll see in the PR, it applies recursively, and can be reapplied at any level, changing how similar structures are matched at different levels of nesting.

**Comment by lennartbuit** [comment]
Yes! This is exactly what I was looking for. 

I agree with the benefits of having a generic `match-with` API over an `match-equals`. We use matcher combinators very much to describe what we are testing, so an intention revealing API is perfect!

I have one question though, there is an assert expressions for `match-equals?`, should that then also be discouraged for the same benefits you list here? Maybe this is not my question to ask, but as a library consumer, it would be good to understand its intended use, right :).

Thank you very much for picking this up, I really appreciate this library, and this discussion in particular!

**Comment by dchelimsky** [comment]
> I have one question though, there is an assert expressions for match-equals?, should that then also be discouraged for the same benefits you list here? Maybe this is not my question to ask, but as a library consumer, it would be good to understand its intended use, right :).

As a library consumer, It is exactly your question to ask!

I hadn't considered it. I can see arguments for and against. Same question could be asked of `match-roughly?`. All of these use the internals of `match-with`.

@philomates WDYT about this?

**Comment by philomates** [comment]
> > I have one question though, there is an assert expressions for match-equals?, should that then also be discouraged for the same benefits you list here? Maybe this is not my question to ask, but as a library consumer, it would be good to understand its intended use, right :).

so like currently we have

```clojure
(defmethod clojure.test/assert-expr 'match-equals? [msg form]
  (matcher-combinators.test/build-match-assert 'match-equals? {clojure.lang.IPersistentMap m/equals} msg form))

(is (match-equals? {:a 1} {:a 1}))
```

and are you suggestion that `clojure.test/assert-expr` mechanics becomes irrelevant? Given you can implement it as

```clojure
(is (match? (match-with {clojure.lang.IPersistentMap m/equals}  {:a 1} {:a 1}))
```
or
```clojure
(def equals-matcher (partial match-with {clojure.lang.IPersistentMap m/equals}))

(is (match? (equals-matcher  {:a 1} {:a 1}))
```

If so, then yes, I agree! I think it would be nice to deprecate all the existing `match-with?` top level forms and have it all reimplemented using this new `match-with` matcher

**Comment by dchelimsky** [comment]
> and are you suggestion that `clojure.test/assert-expr` mechanics becomes irrelevant given you can implement it as the following?
> 
> ```clojure
> (is (match? (match-with {clojure.lang.IPersistentMap m/equals}  {:a 1} {:a 1}))
> ```

Actually, the current version in the PR lets you do this instead:

```clojure
(is (match? (match-with [map? m/equals] {:a 1} {:a 1}))
```

I'm in favor of this change @philomates, and think we should probably do it all in the same release. Agree?

**Comment by philomates** [comment]
@dchelimsky yeah, I think a follow-up PR that deprecates the top level `match-with?` in favor of the new `match-with` matcher sounds great

**Comment by dchelimsky** [comment]
@lennartbuit FYI that we merged #134 but we're doing a couple of other things before releasing. Should be a 3.0 release this week.

**Comment by lennartbuit** [comment]
Really can't stress how positive this interaction was, thank you both very much! We will patiently wait for 3.0 then, we have workarounds in place until that!

<sub>As an anecdote, I recently found a test in our suite that intended to recursively apply `m/equals`, but didn't, so theres at least one test that needs improving ;)</sub>

**Comment by dchelimsky** [comment]
Released in matcher-combinators-3.0.1. Cheers!

**Comment by fernando-nubank** [comment]
Fixed in https://github.com/nubank/matcher-combinators/pull/134, released in https://github.com/nubank/matcher-combinators/releases/tag/3.0.0
